libavfilter/sf_text2graphicsub.c

```diff
@@ -0,0 +1,492 @@
+
+    init_palettizecontext(&context->palettize_context);
+
+    ff_draw_init(&context->draw_context, AV_PIX_FMT_BGRA, FF_DRAW_PROCESS_ALPHA);
+
+    return 0;
+}
+
+static int config_input(AVFilterLink *inlink)
+{
+    const AVFilterContext *ctx = inlink->dst;
+    ASS_Context *context = ctx->priv;
+    if (!context->size.width)  context->size.width  = inlink->w;
+    if (!context->size.height) context->size.height = inlink->h;
+    if (!context->size.height || !context->size.width) {
+        av_log(NULL, AV_LOG_ERROR, "A positive height and width are required to render subtitles\n");
+        return AVERROR_EXIT;
+    }
+    ass_set_frame_size(context->renderer, context->size.width, context->size.height);
+    ass_set_storage_size(context->renderer, context->size.width, context->size.height);
+
+    return 0;
+}
+
+static int config_output(AVFilterLink *outlink)
+{
+    AVFilterContext *ctx = outlink->src;
+    ASS_Context *context = ctx->priv;
+
+    outlink->w = context->size.width;
+    outlink->h = context->size.height;
+
+    return 0;
+}
+
+static int filter_frame(AVFilterLink *inlink, AVFrame *frame)
+{
+    const AVFilterContext *ctx = inlink->dst;
+    ASS_Context *context = ctx->priv;
+    AVFilterLink *outlink = ctx->outputs[0];
+    if (!context->track) {      // Init track
+        if (frame->subtitle_header)
+            context->track = ass_read_memory(context->library,
+                (char *) frame->subtitle_header->data,
+                frame->subtitle_header->size, NULL);
+        else context->track = ass_new_track(context->library);
+        if (!frame->subtitle_header) context->ass_split_context = avpriv_ass_split("");
+        else context->ass_split_context = avpriv_ass_split((char *) frame->subtitle_header->data);
+    }
+
+    for (int r = 0; r < frame->num_subtitle_areas; r++)
+    {
+        ASS_Image *image;
+        ASSDialog *dialog;
+        ASS_Track *track = context->track;
+        AVSubtitleArea *area = frame->subtitle_areas[r];
+        if (area->type == SUBTITLE_BITMAP) continue;
+
+        dialog = avpriv_ass_split_dialog(context->ass_split_context, area->ass);
+        if (track->n_events > 0)
+            ass_free_event(track, 0);
+        track->n_events = 0;
+        ass_alloc_event(track);
+        track->n_events = track->max_events = 1;
+        track->events[0].Start = frame->subtitle_timing.start_pts / (AV_TIME_BASE / 1000);
+        track->events[0].Duration = frame->subtitle_timing.duration / (AV_TIME_BASE / 1000);
+        track->events[0].Effect = av_strdup(dialog->effect);
+        track->events[0].Layer = dialog->layer;
+        track->events[0].MarginL = dialog->margin_l;
+        track->events[0].MarginR = dialog->margin_r;
+        track->events[0].MarginV = dialog->margin_v;
+        track->events[0].Name = av_strdup(dialog->name);
+        track->events[0].Text = av_strdup(dialog->text);
+        track->events[0].ReadOrder = dialog->readorder;
+        track->events[0].Style = 0;
```

Comment:
This doesn't work out well.  avpriv_ass_split_dialog() is much inferior to the original ass parser and I've seen lots of errors and problems with this approach.
Another problem is that you cannot treat each event separate from each other because ass needs to know about all events which are to be shown at a time in order to position them. 
With this approach, you get overlapping text blocks.

Reply:
I don't really see how else I can do it, it's sending it to me one event at a time.

Reply:
You need to feed libass in the same way like the other filters do.

Then it will probably also become more clear why I was talking about splitting the result into multiple drawing areas.

Reply:
Which filter? Give me an example of one with libass.

Reply:
Like overlaytextsubs for example:

https://github.com/softworkz/FFmpeg/blob/d4bb52e6b1c5cb82e829ab75fb07d738e9a1313d/libavfilter/vf_overlaytextsubs.c#L551-L556

There are quite a lot of functions in libass for this (like pars_line, etc.). I don't know which would be best for this case and I don't remember why ass_process_chunk() is preferred in the referenced code.

But what's for sure is that **any** libass function for parsing ass lines is better than the ffmpeg-side ass_split_dialog()

Reply:
All right, I'll look into that.

Reply:
Ok, I did that. The tracks render much better now (if a little slower). Even the animations look (almost) great.